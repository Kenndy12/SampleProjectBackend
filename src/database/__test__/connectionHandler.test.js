const mongoose = require("mongoose");
require("dotenv").config();

jest.useFakeTimers();

describe("connectionHandler test suite", () => {
	afterEach(() => {
		jest.clearAllMocks();
	});

	it("Should have the correct properties", () => {
		const DBConn = require("../connectionHandler");
		expect(DBConn.mongoDB.connect).toBeDefined();
		expect(typeof DBConn.mongoDB.connect).toBe("function");
		expect(DBConn.mongoDB.disconnect).toBeDefined();
		expect(typeof DBConn.mongoDB.disconnect).toBe("function");
	});

	test("When succesfull in making connection", async () => {
		const fakeDBConnString = "fake-dbconnStr";
		const fakeDBName = "fake-dbName";

		process.env.connection_url = fakeDBConnString;
		process.env.db_name = fakeDBName;

		const spyMongoConnect = jest
			.spyOn(mongoose, "connect")
			.mockResolvedValue(true);

		const DBConn = require("../connectionHandler");

		await DBConn.mongoDB.connect();

		expect(spyMongoConnect).toBeCalledTimes(1);
		expect(spyMongoConnect).toBeCalledWith(`${fakeDBConnString}`, {
			useNewUrlParser: true,
			dbName: fakeDBName,
			useUnifiedTopology: true,
		});
	});

	test("When unsuccessfull in making connection", async () => {
		const fakeDBConnString = "fake-dbconnStr";
		const fakeDBName = "fake-dbName";

		process.env.connection_url = fakeDBConnString;
		process.env.db_name = fakeDBName;

		const fakeError = new Error("fake-db-conn-error");
		const spyMongoConnect = jest
			.spyOn(mongoose, "connect")
			.mockResolvedValue(fakeError);

		const DBConn = require("../connectionHandler");

		await DBConn.mongoDB.connect();

		expect(spyMongoConnect).toBeCalledTimes(1);
		expect(spyMongoConnect).toBeCalledWith(`${fakeDBConnString}`, {
			useNewUrlParser: true,
			dbName: fakeDBName,
			useUnifiedTopology: true,
		});
	});
});

describe("Check registered event handers", () => {
	test("On connected", () => {
		const onConnected = mongoose.connection._events["connected"];
		expect(typeof onConnected).toBe("function");
		const spyConsoleLog = jest.spyOn(console, "log");

		const result = onConnected();
		expect(spyConsoleLog).toHaveBeenCalledTimes(1);
		expect(spyConsoleLog.mock.calls).toMatchSnapshot();

		expect(result).toBe(undefined);
	});

	test("on reconnected", () => {
		const onReconnected = mongoose.connection._events.reconnected;

		expect(typeof onReconnected).toBe("function");
		const spyConsoleLog = jest.spyOn(console, "log");

		const result = onReconnected();

		expect(spyConsoleLog).toHaveBeenCalledTimes(1);
		expect(spyConsoleLog.mock.calls).toMatchSnapshot();

		expect(result).toBe(undefined);
	});
});
