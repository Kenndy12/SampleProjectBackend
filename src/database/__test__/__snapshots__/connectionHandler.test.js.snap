// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check registered event handers On connected 1`] = `
[
  [
    "DB Connected",
  ],
]
`;

exports[`Check registered event handers on reconnected 1`] = `
[
  [
    "DB Reconnected",
  ],
]
`;
